File: src/main/scala/uk/co/ridentbyte/view/response/BulkRequestOutputPane.scala
package uk.co.ridentbyte.view.response

import javafx.geometry.Insets
import javafx.scene.chart.{AreaChart, NumberAxis, XYChart}
import javafx.scene.control.{Button, TextArea}
import javafx.scene.layout.{GridPane, Priority}
import uk.co.ridentbyte.model.{CardinalRequest, CardinalRequestAndResponse}
import uk.co.ridentbyte.view.util.RowConstraintsBuilder
import scala.collection.JavaConverters._

case class BulkRequestOutputPane(requestAndResponses: List[CardinalRequestAndResponse],
                                 exportToCsv: java.util.function.Function[java.util.List[CardinalRequestAndResponse], Void],
                                 exportToBash: java.util.function.BiFunction[java.util.List[CardinalRequest], java.lang.Integer, Void],
                                 throttle: Int) extends GridPane {

  val responses = requestAndResponses.map(_.getResponse).filter(_ != null)

  setHgap(10)
  setVgap(10)
  setPadding(new Insets(20))
  getStyleClass.addAll("plain-border", "round-border")

  getRowConstraints.addAll(
    new RowConstraintsBuilder().withPercentageHeight(55).withVgrow(Priority.ALWAYS).build,
    new RowConstraintsBuilder().withVgrow(Priority.ALWAYS).build,
    new RowConstraintsBuilder().withVgrow(Priority.NEVER).build
  )

  val xAxis = new NumberAxis
  xAxis.setForceZeroInRange(true)
  xAxis.setMinorTickVisible(false)
  xAxis.setTickLabelsVisible(false)

  val yAxis = new NumberAxis
  yAxis.setTickLabelFormatter(new NumberAxis.DefaultFormatter(yAxis, null, " ms"))
  yAxis.setForceZeroInRange(true)

  val timeSeries = new XYChart.Series[Number, Number]
  responses.zipWithIndex.foreach { case(r, i) =>
    timeSeries.getData.add(new XYChart.Data(i, r.getTime))
  }

  val lineChart = new AreaChart[Number, Number](xAxis, yAxis)
  lineChart.setTitle("Response time over time")
  lineChart.setCreateSymbols(false)
  lineChart.setLegendVisible(false)
  lineChart.getData.add(timeSeries)
  GridPane.setHgrow(lineChart, Priority.ALWAYS)
  GridPane.setColumnSpan(lineChart, 2)
  add(lineChart, 0, 0)

  val allTimes = responses.map(_.getTime)
  val allTimesSorted = allTimes.sorted
  val averageResponseTime = if (responses.isEmpty) 0 else allTimes.sum / responses.size

  val responseCodesWithCounts = responses.groupBy(_.getStatusCode)

  val requestCountsOutput = if (responseCodesWithCounts.isEmpty) {
    List("No responses")
  } else {
    responseCodesWithCounts.map { case (c, r) =>
      s"HTTP $c.................. ${r.size}"
    }
  }

  val timings =
    s"""Timings
       |---
       |Average Response Time..... $averageResponseTime ms
       |Fastest Response Time..... ${allTimesSorted.headOption.getOrElse(0)} ms
       |Slowest Response Time..... ${allTimesSorted.lastOption.getOrElse(0)} ms
       |
         |Request/Response Counts
       |---
       |${requestCountsOutput.mkString("\n")}""".stripMargin

  val textAreaTimings = new TextArea(timings)
  textAreaTimings.getStyleClass.add("cardinal-font-console")
  GridPane.setHgrow(textAreaTimings, Priority.ALWAYS)
  GridPane.setColumnSpan(textAreaTimings, 2)
  add(textAreaTimings, 0, 1)

  val exportToCsvButton = new Button("Export to CSV...")
  exportToCsvButton.setOnAction(_ => exportToCsv.apply(requestAndResponses.asJava))
  GridPane.setColumnSpan(exportToCsvButton, 1)
  GridPane.setHgrow(exportToCsvButton, Priority.NEVER)
  add(exportToCsvButton, 0, 2)

  val exportToBashButton = new Button("Export as script...")
  exportToBashButton.setOnAction(_ => exportToBash.apply(requestAndResponses.map(_.getRequest).asJava, throttle))
  GridPane.setColumnSpan(exportToBashButton, 1)
  GridPane.setHgrow(exportToBashButton, Priority.ALWAYS)
  add(exportToBashButton, 1, 2)
}

File: src/main/scala/uk/co/ridentbyte/view/response/BulkRequestOutputPane.java
package uk.co.ridentbyte.view.response;

import javafx.geometry.Insets;
import javafx.scene.chart.AreaChart;
import javafx.scene.chart.NumberAxis;
import javafx.scene.chart.XYChart;
import javafx.scene.control.Button;
import javafx.scene.control.TextArea;
import javafx.scene.layout.GridPane;
import javafx.scene.layout.Priority;
import uk.co.ridentbyte.model.CardinalRequest;
import uk.co.ridentbyte.model.CardinalRequestAndResponse;
import uk.co.ridentbyte.model.CardinalResponse;
import uk.co.ridentbyte.view.util.RowConstraintsBuilder;

import java.util.List;
import java.util.Map;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.stream.Collectors;

public class BulkRequestOutputPane extends GridPane {

    public BulkRequestOutputPane(List<CardinalRequestAndResponse> requestAndResponses,
                                 Function<List<CardinalRequestAndResponse>, Void> exportToCsv,
                                 BiFunction<List<CardinalRequest>, Integer, Void> exportToBash,
                                 int throttle) {
        this.setHgap(10);
        this.setVgap(10);
        this.setPadding(new Insets(20));
        this.getStyleClass().addAll("plain-border", "round-border");
        this.getRowConstraints().addAll(
                new RowConstraintsBuilder().withPercentageHeight(55).withVgrow(Priority.ALWAYS).build(),
                new RowConstraintsBuilder().withVgrow(Priority.ALWAYS).build(),
                new RowConstraintsBuilder().withVgrow(Priority.NEVER).build()
        );

        List<CardinalResponse> responses = requestAndResponses.stream()
                .map(CardinalRequestAndResponse::getResponse)
                .collect(Collectors.toList());

        List<CardinalRequest> requests = requestAndResponses.stream()
                .map(CardinalRequestAndResponse::getRequest)
                .collect(Collectors.toList());

        NumberAxis xAxis = new NumberAxis();
        xAxis.setForceZeroInRange(true);
        xAxis.setMinorTickVisible(false);
        xAxis.setTickLabelsVisible(false);

        NumberAxis yAxis = new NumberAxis();
        yAxis.setTickLabelFormatter(new NumberAxis.DefaultFormatter(yAxis, null, " ms"));
        yAxis.setForceZeroInRange(true);

        XYChart.Series<Number, Number> timeSeries = new XYChart.Series<>();
        for (int i = 0; i < responses.size(); i++) {
            timeSeries.getData().add(new XYChart.Data<>(i, responses.get(i).getTime()));
        }

        AreaChart<Number, Number> lineChart = new AreaChart<>(xAxis, yAxis);
        lineChart.setTitle("Response time over time");
        lineChart.setCreateSymbols(false);
        lineChart.setLegendVisible(false);
        lineChart.getData().add(timeSeries);
        GridPane.setHgrow(lineChart, Priority.ALWAYS);
        GridPane.setColumnSpan(lineChart, 2);
        add(lineChart, 0, 0);


        List<Long> allResponseTimes = responses.stream()
                .map(CardinalResponse::getTime)
                .collect(Collectors.toList());

        List<Long> allResponseTimesSorted = allResponseTimes.stream().sorted().collect(Collectors.toList());

        long averageResponseTime = 0;
        if (!responses.isEmpty()) {
            averageResponseTime = allResponseTimes.stream().mapToLong(Long::longValue).sum() / responses.size();
        }

        Map<Integer, List<CardinalResponse>> responsesGroupedByStatusCode =
                responses.stream().collect(Collectors.groupingBy(CardinalResponse::getStatusCode));

        StringBuilder timingsOutput = new StringBuilder();
        timingsOutput.append("Timings\n");
        timingsOutput.append("---\n");
        timingsOutput.append("Average Response Time..... ");
        timingsOutput.append(averageResponseTime);
        timingsOutput.append("ms\n");
        timingsOutput.append("Fastest Response Time..... ");
        timingsOutput.append(allResponseTimesSorted.size() > 0 ? allResponseTimesSorted.get(0) : "0");
        timingsOutput.append("ms\n");
        timingsOutput.append("Slowest Response Time..... ");
        timingsOutput.append(allResponseTimesSorted.size() > 0 ? allResponseTimesSorted.get(allResponseTimesSorted.size() - 1) : "0");
        timingsOutput.append("ms\n");
        timingsOutput.append("\n");
        timingsOutput.append("Request/Response Counts\n");
        timingsOutput.append("---\n");
        if (responsesGroupedByStatusCode.isEmpty()) {
            timingsOutput.append("No responses");
        } else {
            for (Map.Entry<Integer, List<CardinalResponse>> statusWithResponses : responsesGroupedByStatusCode.entrySet()) {
                timingsOutput.append("Http ");
                timingsOutput.append(statusWithResponses.getKey());
                timingsOutput.append(".................. ");
                timingsOutput.append(statusWithResponses.getValue().size());
                timingsOutput.append("\n");
            }
        }

        TextArea textAreaTimings = new TextArea(timingsOutput.toString());
        textAreaTimings.getStyleClass().add("cardinal-font-console");
        GridPane.setHgrow(textAreaTimings, Priority.ALWAYS);
        GridPane.setColumnSpan(textAreaTimings, 2);
        add(textAreaTimings, 0, 1);

        Button exportToCsvButton = new Button("Export to CSV...");
        exportToCsvButton.setOnAction((a) -> exportToCsv.apply(requestAndResponses));
        GridPane.setColumnSpan(exportToCsvButton, 1);
        GridPane.setHgrow(exportToCsvButton, Priority.NEVER);
        add(exportToCsvButton, 0, 2);

        Button exportToBashButton = new Button("Export as script...");
        exportToBashButton.setOnAction((a) -> exportToBash.apply(requests, throttle));
        GridPane.setColumnSpan(exportToBashButton, 1);
        GridPane.setHgrow(exportToBashButton, Priority.ALWAYS);
        add(exportToBashButton, 1, 2);

    }

}


